### PR TITLE
added data mapper [WIP]

### PR DESCRIPTION
**Pull request by dg**
- bug fix? no 
- new feature? yes
- BC break? no

Added built-in support for mapping form values to object and vice versa.

### CODE REVIEW

File: src/Forms/Container.php [hunk 1]
```diff
@@ -56,53 +66,88 @@ public function setDefaults(iterable $values, bool $erase = false)
...
-		$values = $asArray ? [] : new Nette\Utils\ArrayHash;
+		$mappedClass = $asArray ? null : $this->mappedClass;
+		$mapper = $this->mapper ?: new Mapper;
+		$values = [];
 		foreach ($this->getComponents() as $name => $control) {
 			if ($control instanceof IControl && !$control->isOmitted()) {
 				$values[$name] = $control->getValue();
 
 			} elseif ($control instanceof self) {
+				if ($mappedClass && $control->mappedClass === self::DEFAULT_MAPPED_CLASS) {
```
Comment (milo) [inline src/Forms/Container.php]: For ArrayHash only to keep the compatibility? Or is there another reason?

Comment (dg) [reply src/Forms/Container.php]: I want to keep current functionality.

File: src/Forms/Container.php [hunk 2]
```diff
@@ -56,53 +66,88 @@ public function setDefaults(iterable $values, bool $erase = false)
...
 
-			} elseif ($control instanceof self) {
-				if (array_key_exists($name, $values)) {
-					$control->setValues($values[$name], $erase);
+		foreach ($components as $name => $control) {
+			if (array_key_exists($name, $data)) {
+				if ($control instanceof IControl) {
+					$control->setValue($data[$name]);
+				} else {
+					$control->setValues($data[$name] ?: [], $erase);
```
Comment (JanTvrdik) [inline src/Forms/Container.php]: The choice to use `?:` instead of `??` is intentional?

File: src/Forms/Container.php [hunk 3]
```diff
@@ -56,53 +66,88 @@ public function setDefaults(iterable $values, bool $erase = false)
...
+				}
 
-				} elseif ($erase) {
+			} elseif ($erase) {
+				if ($control instanceof IControl) {
```
Comment (JanTvrdik) [inline src/Forms/Container.php]: I don't like those conditions. How about we split `$components` array into `$controls`, `$containers` and `$dataKeys`.

File: src/Forms/Container.php [hunk 4]
```diff
@@ -56,53 +66,88 @@ public function setDefaults(iterable $values, bool $erase = false)
...
 	/**
 	 * Returns the values submitted by the form.
-	 * @return Nette\Utils\ArrayHash|array
+	 * @return object|array
 	 */
 	public function getValues(bool $asArray = false)
 	{
-		$values = $asArray ? [] : new Nette\Utils\ArrayHash;
+		$mappedClass = $asArray ? null : $this->mappedClass;
+		$mapper = $this->mapper ?: new Mapper;
```
Comment (JanTvrdik) [inline src/Forms/Container.php]: The choice to use `?:` instead of `??` is intentional?

File: src/Forms/Container.php [hunk 5]
```diff
@@ -33,6 +35,12 @@ class Container extends Nette\ComponentModel\Container implements \ArrayAccess
 	/** @var bool */
 	private $validated;
 
+	/** @var string|null */
+	private $mappedClass = self::DEFAULT_MAPPED_CLASS;
+
+	/** @var IMapper */
```
Comment (JanTvrdik) [inline src/Forms/Container.php]: `IMapper|null`

**Pull request closed without merging**